### PR TITLE
Ev/cdr bug fix

### PR DESCRIPTION
**Pull request by eyalvo-matrix**
Hi to all , 
Reviewing the CDR engine  I have noticed that in case of a form that is being created and deleted 
the CDR still consider the form as created.
I added a code to fix that. 
The code apply only to standard forms. 
Assuming "form_ " + name format .

### CODE REVIEW

**Comment by mdsupport** [comment]
[PR 359](https://github.com/openemr/openemr/pull/359/commits/5a9aac9fa6ad59800133ae064898b3cddb96f5ab) provides a generic approach to configuring "non standard" tables - those without "form_" prefix, "id" as primary key or 1:1 records. 

Until an encounter object is put in place or project starts using sql views, bugs will rule the codebase.

**Comment by eyalvo-matrix** [comment]
Hi fellow developer,

Thank you for your replay.
I fear I did not fully understand your answer .
My fix is indeed specific for form_  prefix .
I understand that in the future or even now forms can be configured in a
different way.
We currently do need this fix since our OpenEmr is already in use .
I attached an example of our Db
Can you please tell me what approach should we take to solve the problem?
1:1 id is currently not an option for my best understanding.

thanks in advance ,

Eyal

File: library/clinical_rules.php [hunk 1]
```diff
@@ -1914,13 +1914,28 @@ function exist_database_item($patient_id,$table,$column='',$data_comp,$data='',$
...
   }
   else {
-    // search for number of specific items
-    $sql = sqlStatementCdrEngine("SELECT `" . add_escape_custom($column) . "` " .
-      "FROM `" . add_escape_custom($table)  . "` " .
-      " ". $whereTables. " ".
-      "WHERE `" . add_escape_custom($column) ."`" . $compSql . "? " .
-      "AND " . add_escape_custom($patient_id_label)  . "=? " . $customSQL .
-      $dateSql, array($data,$patient_id) );
+      if ($whereTables=="" && strpos($table, 'form_')!== false){
```
Comment (bradymiller) [inline library/clinical_rules.php]: Do you also need to support when a column is not given?

File: library/clinical_rules.php [hunk 2]
```diff
@@ -1914,13 +1914,28 @@ function exist_database_item($patient_id,$table,$column='',$data_comp,$data='',$
...
+          //To handle standard forms starting with form_
+          $sql =sqlStatementCdrEngine(
+              "SELECT b.`" . add_escape_custom($column) . "` " .
+              "FROM forms a ".
+              "LEFT JOIN `" . add_escape_custom($table) . "` " . " b ".
+              "ON (a.form_id=b.id AND a.formdir LIKE '".add_escape_custom(substr($table, 5))."')".
+              "WHERE a.deleted != '1'".
+              "AND b.`" .add_escape_custom($column) ."`" . $compSql . "? " .
+              "AND b."  . add_escape_custom($patient_id_label)  . "=? " . $customSQL
+              .substr_replace($dateSql,'b.',5,0)
```
Comment (bradymiller) [inline library/clinical_rules.php]: Was hoping to get clarification on what this substr_replace of the $dateSql is doing.

**Comment by bradymiller** [comment]
Posted some questions/comments in the code.(minor issues)
thanks, -brady

File: library/clinical_rules.php (hunk 2 above)
Comment (eyalvo-matrix) [inline library/clinical_rules.php]: The substr_replace intended to change the $dateSql
 from 
AND (`date` BETWEEN DATE_SUB('......
to  
AND (b.`date` BETWEEN DATE_SUB('......
adding "b." to date 
I did that here and not in the creation of  $dateSql in order to leave the else case unchanged

File: library/clinical_rules.php (hunk 1 above)
Comment (eyalvo-matrix) [inline library/clinical_rules.php]: No , this handles only the specific case that a column is given 
and  that the table represent a form

File: library/clinical_rules.php (hunk 2 above)
Comment (eyalvo-matrix) [inline library/clinical_rules.php]: Before the change the $dateSql contained something like this :
AND (`date` BETWEEN DATE_SUB('2016-12-28 11:53:37', INTERVAL 6 MONTH) AND '2016-12-28 11:53:37')

**Comment by mdsupport** [comment]
@eyalvo-matrix,
exist_database_item seemed to be scouring the database for all patient data.  Since there are existing variations in encounter ->> forms access, I want to suggest making PR 359 type of approach reusable and then build the list of exceptions rather than relying on hard coded table names put in by other developers.
Cheers.

**Comment by eyalvo-matrix** [comment]
Ok , I will review the PR 359 in depth and try to re-post the code here . 
It will take few days. If anyone else wants to take the gloves feel free :)

File: library/clinical_rules.php (hunk 2 above)
Comment (bradymiller) [inline library/clinical_rules.php]: thanks for the explanation. makes good sense.

**Comment by bradymiller** [comment]
@eyalvo-matrix and @mdsupport ,

I'd consider using mdsupport's proposed approach optional (it would, however, be a extremely helpful feature in the long run).

Since your goal was to simply not include deleted forms, I think your current code submission would be acceptable to go in the codebase as long as it tests well (just please ensure that the rules that test the vitals signs are working with your changes).

**Comment by eyalvo-matrix** [comment]
Hi Brady I have applied the fixes can you please review the code ?
Thanks a lot ! Eyal .

**Comment by bradymiller** [comment]
Hi @eyalvo-matrix ,

Do you mean this PR or your other PR here:
https://github.com/openemr/openemr/pull/373

For this PR, I just need to know that your testing with the vitals rules are still working with your changes (and then I'd bring this PR in)

For the other PR, note my comment on changing the prices_history table to something like codes_history table.

thanks, -brady

**Comment by eyalvo-matrix** [comment]
Hi

As far as I checked For this PR - vitals rules indeed still working though
I do not know how to write and check complex rules

About the #373 <https://github.com/openemr/openemr/pull/373> I refactored
the code and changed it to code history as you instructed

about the reports we indeed planning (Not official yet ) to integrate
Openemr with Pentaho BI .

**Comment by bradymiller** [comment]
just brought this into the codebase. thanks for the fix!

**Comment by eyalvo-matrix** [comment]
Thanks :)

**Comment by bradymiller** [comment]
FYI,

There were reports of the following error happening (not on all systems):

**Comment by bradymiller** [comment]
oops, will try again:

FYI,

There were reports of the following error happening (not on all systems):


I committed a fix for this to the codebase:
https://github.com/openemr/openemr/commit/19664124ee7fe8203fd6f3989d8778214728fe56

**Comment by bradymiller** [comment]
To be clear, I committed a fix for this here:
https://github.com/openemr/openemr/commit/19664124ee7fe8203fd6f3989d8778214728fe56

**Comment by eyalvo-matrix** [comment]
thanks :)

**Pull request merged**